packages/react-reconciler/src/ReactFiberCompleteWork.js

```diff
@@ -930,112 +1004,160 @@ function completeWork(
-              // them, then they really have the same priority as this render.
-              // So we'll pick it back up the very next render pass once we've had
-              // an opportunity to yield for paint.
+          // hasSuspendedChildrenAndNewContent could've set didSuspendAlready
+          didSuspendAlready =
+            (workInProgress.effectTag & DidCapture) !== NoEffect;
+        }
+        if (didSuspendAlready) {
+          cutOffTailIfNeeded(renderState);
+        }
+        // Next we're going to render the tail.
+      } else {
+        // Append the rendered row to the child list.
+        if (!didSuspendAlready) {
+          if (
+            now() > renderState.tailExpiration &&
+            renderExpirationTime > Never
+          ) {
+            // We have now passed our CPU deadline and we'll just give up further
+            // attempts to render the main content and only render fallbacks.
+            // The assumption is that this is usually faster.
+            workInProgress.effectTag |= DidCapture;
+            didSuspendAlready = true;
 
-              const nextPriority = renderExpirationTime - 1;
-              workInProgress.expirationTime = workInProgress.childExpirationTime = nextPriority;
-              if (enableSchedulerTracing) {
-                markSpawnedWork(nextPriority);
-              }
-            } else {
-              suspenseListState.didSuspend = isShowingAnyFallbacks(rendered);
+            cutOffTailIfNeeded(renderState);
+
+            // Since nothing actually suspended, there will nothing to ping this
+            // to get it started back up to attempt the next item. If we can show
+            // them, then they really have the same priority as this render.
+            // So we'll pick it back up the very next render pass once we've had
+            // an opportunity to yield for paint.
+
+            const nextPriority = renderExpirationTime - 1;
+            workInProgress.expirationTime = workInProgress.childExpirationTime = nextPriority;
+            if (enableSchedulerTracing) {
+              markSpawnedWork(nextPriority);
+            }
+          } else if (isShowingAnyFallbacks(renderedTail)) {
+            workInProgress.effectTag |= DidCapture;
+            didSuspendAlready = true;
+            cutOffTailIfNeeded(renderState);
+            // This might have been modified.
+            if (renderState.rendering === null) {
+              // Ensure we transfer the update queue to the parent.
+              // TODO: This just reuses the code from the other path but could
+              // be optimized better.
+              hasSuspendedChildrenAndNewContent(workInProgress, renderedTail);
+              // Reset the effect list to what it w as before we rendered this
+              // child. The nested children have already appended themselves.
+              workInProgress.lastEffect = renderState.lastEffect;
```

Comment:
We probably need to set `workInProgress.lastEffect.nextEffect = null;`. Most likely this will get set anyway but just to be safe. E.g. the suspense list its gets added to the end of the list.